src/algebra/category/Group/epi_mono.lean

```diff
@@ -0,0 +1,334 @@
+
+lemma h_apply_from_coset' (x : B) (b : B) (hb : b ∈ f.range):
+  (h x) (from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩) =
+  from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩ :=
+(from_coset_eq_of_mem_range _ hb).symm ▸ h_apply_from_coset f x
+
+lemma h_apply_from_coset_nin_range (x : B) (hx : x ∈ f.range) (b : B) (hb : b ∉ f.range) :
+  (h x) (from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩) =
+  from_coset ⟨(x * b) *l f.range.carrier, ⟨x * b, rfl⟩⟩ :=
+begin
+  simp only [H, tau, monoid_hom.coe_mk, equiv.to_fun_as_coe, equiv.coe_trans, function.comp_app],
+  rw [equiv.symm_swap],
+  rw @equiv.swap_apply_of_ne_of_ne X' _
+    (from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩) ∞
+    (from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩) (from_coset_ne_of_nin_range _ hb) (by simp),
+  simp only [g_apply_from_coset, ←subtype.val_eq_coe, left_coset_assoc],
+  exact equiv.swap_apply_of_ne_of_ne begin
+    refine from_coset_ne_of_nin_range _ (λ r, hb _),
+    convert subgroup.mul_mem _ (subgroup.inv_mem _ hx) r,
+    rw [←mul_assoc, mul_left_inv, one_mul],
+  end (by simp),
+end
+
+lemma agree :
+  f.range.carrier = {x | h x = g x} :=
+set.ext $ λ b,
+⟨begin
+  rintros ⟨a, rfl⟩,
+  change h (f a) = g (f a),
+  ext ⟨⟨_, ⟨y, rfl⟩⟩⟩,
+  { rw [g_apply_from_coset],
+    by_cases m : y ∈ f.range,
+    { rw [h_apply_from_coset' _ _ _ m, from_coset_eq_of_mem_range _ m],
+      change from_coset _ = from_coset ⟨f a *l (y *l _), _⟩,
+      simpa only [←from_coset_eq_of_mem_range _ (subgroup.mul_mem _ ⟨a, rfl⟩ m),
+        left_coset_assoc] },
+    { rw [h_apply_from_coset_nin_range _ _ ⟨_, rfl⟩ _ m],
+      simpa only [←subtype.val_eq_coe, left_coset_assoc], }, },
+  { rw [g_apply_infinity, h_apply_infinity _ _ ⟨_, rfl⟩], },
+end, λ (hb : h b = g b), classical.by_contradiction $ λ r, begin
+  have eq1 : (h b) (from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩) =
+    (from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩) := by simp [H, tau, g_apply_infinity],
+  have eq2 : (g b) (from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩) =
+    (from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩) := rfl,
+  exact (from_coset_ne_of_nin_range _ r).symm (by rw [←eq1, ←eq2, fun_like.congr_fun hb]),
+end⟩
+
+lemma comp_eq : f ≫ (show B ⟶ Group.of SX', from g) = f ≫ h :=
+fun_like.ext _ _ $ λ a,
+  by simp only [comp_apply, show h (f a) = _, from (by simp [←agree] : f a ∈ {b | h b = g b})]
+
+lemma g_ne_h (x : B) (hx : x ∉ f.range) :
+  g ≠ h :=
+λ r, begin
+  replace r := fun_like.congr_fun (fun_like.congr_fun r x)
+    ((from_coset ⟨f.range, ⟨1, one_left_coset _⟩⟩)),
+  rw [H, g_apply_from_coset, monoid_hom.coe_mk, tau] at r,
+  simp only [monoid_hom.coe_range, subtype.coe_mk, equiv.symm_swap,
+    equiv.to_fun_as_coe, equiv.coe_trans, function.comp_app] at r,
+  erw [equiv.swap_apply_left, g_apply_infinity, equiv.swap_apply_right] at r,
+  exact from_coset_ne_of_nin_range _ hx r,
+end
+
+end surjective_of_epi_auxs
+
+lemma surjective_of_epi [epi f] : function.surjective f :=
```

Comment:
I think it's really important that we get this also for `AddGroup` (and also for `CommGroup` and `AddCommGroup`, but that can wait). If `to_additive` really can't work, we could construct an equivalence between `Group` and `AddGroup` and use that it commutes with the forgetful functors and deduce from that that the forgetful functor from `AddGroup`  preserves epimorphisms. But I guess it's fine to wait with this until a follow-up PR.

Reply:
`to_additive` doesn't work, your suggestion is exactly my plan

Reply:
I have opened #15496